lib/control-plane/openstackcontrolplane.yaml

```diff
@@ -164,16 +174,19 @@ spec:
           dbType: NB
           networkAttachment: internalapi
           storageRequest: 10G
+          replicas: 3
         ovndbcluster-sb:
           dbType: SB
           networkAttachment: internalapi
           storageRequest: 10G
+          replicas: 3
```

Comment:
of we are runing the ovn north and south db in a 3 replica cluster we should likely also do the same for OVNNorthd which is the deamon the process the db changes and responds to neutorn vai the websocket.

Reply:
bumped northd to 3

Reply:
There is no relationship between `northd` and OVN RAFT members number, let me explain.

---

`northd` is active-passive HA. It grabs a OVS database lock, then connects to RAFT leaders (for NB and SB) and receives changes in NB and SB databases and translate between the two. The other replicas are just sitting there, waiting for the "main" `northd` to exit - at which point they will try to grab the lock and take over. In the meantime, each `northd` copy maintains its own copy of OVN IDL state in-memory - and connections to RAFT leaders - this has obvious load impact (memory, CPU - incl. for RAFT leaders that have to replicate changes to each "redundant" `northd` instance...)

I assume that, in general, `kubelet` should respawn the container with `northd` if it ever dies. It may take some time to respawn the process, but that's probably OK since `northd` works asynchronously to user visible operations, like Neutron API requests - or OVN database changes coming from Neutron agents or `ovn-controller`s.

Perhaps native `northd` active-passive HA implementation is more robust, in that some failure scenarios for `northd` may not be properly caught by `kubelet` and the container liveness checks. At the moment the check is just `pidof northd`, which is not ideal. (There's an `ovs-appctl status` command available for the service that would probably be a better way to confirm `northd` is not dead-locked.)

(The topic of inadequate liveness checks both in OCP and EDP containers was discussed before but nothing came out of it so far. I will report an issue to improve it in the next few days.)

Perhaps in some situations, having 2 `northd` instances may be a bit more robust, and HA takeover may happen quicker than container restart. (But see above: why do we care if NB changes are translated into SB in a second and not in 10 seconds **during disaster recovery / evacuation** that is an exceptional scenario?)

As I said, an argument **can** be made for bumping it to 2 (to keep the second passive replica). (Though I'd say we should have data to show the benefit is worth it.) But bumping to 3 is just excessive.

(Side comment: I don't think `neutron` ever talks to `northd` directly, via websocket or otherwise.)

Adding @otherwiseguy to keep me honest.

Reply:
ack i was not aware that it was internally looking an thought it could scale horizontally since that is not the case i would suggest keeping this at one or at most 2 for ha reason but we were originally dining this for performance reasons.
it sounds like this will not improve performance so if we want to do this for ha reasons we may want to drop this from the patch and consider that separately.